proxy-server/src/main/java/io/micrometer/prometheus/rsocket/PrometheusControllerProperties.java

```diff
@@ -48,4 +57,16 @@ public int getWebsocketPort() {
   public void setWebsocketPort(int websocketPort) {
     this.websocketPort = websocketPort;
   }
+
+  public Duration getTimeoutOffset() {
```

Comment:
Why is this needed? Should not we set it to whatever value the client wants?

Reply:
If the timeout was set to the exact value of the Prometheus scrape timeout (given by the header), any *one* app instance that is too slow to respond (thus timing out), would also cause us to hit the Prometheus scrape timeout (since they're the same in this hypothetical). Thus, as a result of a *single* instance timing out, no metrics are successfully scraped from *any* of the other connected instances (because Prometheus hit its scrape timeout and closed the connection).

So, a small offset is needed to make the individual instance scrape timeout slightly less than the Prometheus scrape timeout, allowing time for the metrics to be scraped from all connected instances, concatenated and returned via the network before Prometheus itself times out.

Reply:
If this is needed for the reason you describe, shouldn't we set the default to a small but nonzero value?

Reply:
It depends. Defaulting to a zero offset maintains backwards compatibility with the old behavior but doesn't give the benefit of the offset from a timeout perspective.

On the other hand, defaulting to a nonzero offset gives the timeout benefit described above but technically is a breaking change. For example, consider a Prometheus setup with a 10s scrape timeout and assume we chose an offset of 1s. In this case, the RSocket timeout is 9s. Let's say the user running this Prometheus setup previously had RSocket scrape duration of about 9.5 seconds. Prior to this change, they'd be running their setup without issue since there's no explicit RSocket timeout and 9.5 seconds is lower than the Prometheus 10s timeout. However, with the introduction of this change, suddenly this user's scrapes are timing out on the RSocket side from the 9 second timeout and thus aren't getting the metrics they were getting previously.

I don't think this is necessarily a likely scenario and the odds of this happening decrease as the default offset also decreases, but it's definitely a consideration.

Thoughts?

Reply:
I really appreciate you thinking about backward compatibility and the potential negative effects of this. It's important for us to take these things seriously and avoid as much pain for users as possible. That said, I think if we're changing this behavior in a minor release (as opposed to a patch release), we make note of it in the release notes, and it is configurable so people negatively affected by it can effectively undo the behavior change, I think it is worthwhile moving things in a better direction. That's how I feel anyway. @jonatan-ivanov What do you think?

Reply:
I looked into this, it makes sense, Prometheus sends the `scrape_timeout` value in `X-Prometheus-Scrape-Timeout-Seconds` and I did not find a way to account for latency on the Prometheus server side. I find this unfortunate since:

1. If you have multiple Prometheus servers in different network locations with different latencies (let's pretend it is significant enough), the scraped app needs to account for the higher network latency. This feels somewhat wrong because the scraped app should not be aware of this.
2. Or if you move a Prometheus server to one network location to another (let's pretend the latency it is significant enough), you need to change the offset for every scraped app.

But if the offset would be on Prometheus side the issue is somewhat similar:
1. Scraped apps can be in multiple network locations so Prometheus should account for highest network latency, though configuring this per job could solve this.
2. Or if you move an app that is scraped, you need to change all Prometheus instances, though you usually have less Prometheus instances than apps.

So I guess this is what we have, it makes sense to me.

I think I agree with @shakuzen about having a small but non-zero default value in a milestone release. To me the interesting part is what should that value be.
I should be small enough so the negative effect is minimal and big enough so that it can account the network delay which includes transfer cost which depends on the response size. My first guess is around 100-500ms.

Reply:
What do you think @jdlafaye about having a small default offset?

Reply:
I think it's reasonable to set a small default offset as it will provide an immediate benefit in a number of setups. I'd probably lean towards 500ms as that meshes well with the default Prometheus scrape timeout of 10s and also happens to be the [same as what Blackbox Exporter uses](https://github.com/prometheus/blackbox_exporter/blob/master/main.go#L52).

Let me know your thoughts and I can make that change.